iocmanager/config.py

```diff
@@ -0,0 +1,591 @@
+            for k in list(kw.items()):
+                if getattr(ioc, k[0]) != k[1]:
+                    break
+            else:
+                configs.append([cfg, ioc])
+                pass
+    return configs
+
+
+def get_hutch_list() -> list[str]:
+    """Return the list of all supported hutches."""
+    try:
+        config_paths = Path(env_paths.CONFIG_DIR).glob("*/iocmanager.cfg")
+        return [pth.parent.name for pth in config_paths]
+    except Exception:
+        return []
+
+
+def validate_config(config: Config) -> bool:
+    """
+    Returns True if the configuration looks valid.
+
+    Currently, just checks if there is a duplicate host/port combination.
+    """
+    procs = list(config.procs.values())
+    for idx in range(len(procs)):
+        host1 = procs[idx].host
+        port1 = procs[idx].port
+        for jdx in range(idx + 1, len(procs)):
+            host2 = procs[jdx].host
+            port2 = procs[jdx].port
+            if host1 == host2 and port1 == port2:
+                return False
+    #
+    # Anything else we want to check here?!?
+    #
+    return True
+
+
+@dataclass(eq=True)
+class IOCStatusFile:
+    """Information about the IOC based on the IOC status file."""
+
+    name: str
+    port: int
+    host: str
+    path: str
+    pid: int
+    mtime: float = 0.0
+
+    def get_file_location(self, hutch: str) -> str:
+        """
+        Return the filepath to where this status file came from.
+
+        Parameters
+        ----------
+        hutch : str
+            The hutch that owns this IOC.
+        """
+        return env_paths.STATUS_DIR % hutch + "/" + self.name
+
+
+def read_status_dir(cfg: str) -> list[IOCStatusFile]:
+    """
+    Update a status directory for a hutch and return its information.
+
+    This can remove outdated info in the status directory, but it
+    will not add new information. The new information is provided
+    by start_proc.
+
+    Each hutch has a status directory, nominally at
+    /cds/group/pcds/pyps/config/.status/$hutchname
+
+    This directory contains one file per IOC process, which stores:
+    - PID
+    - hostname
+    - procServ port
+    - path to IOC
+
+    The file stores this info in a single line, for example:
+    14509 ctl-tmo-misc-01 30305 ioc/tmo/pvNotepad/R1.1.5
+
+    This function will open each of these files and do the following:
+    - If the file doesn't have 4 parts, delete the file
+    - If we encounter multiple files with the same host/port combination,
+      delete all but the newest such file.
+    - Collect information about the files that remain and return it all
+
+    Parameters
+    ----------
+    cfg : str
+        The hutch name associated with the config, such as xpp or tmo.
+
+    Returns
+    -------
+    status : list of IOCStatus
+        A list of structured data containing all information about each
+        IOC from the status dir.
+    """
+    info: dict[tuple[str, str], IOCStatusFile] = {}
```

Comment:
It might be worth a type alias for the host-port combination, or at least a comment up top that this is what the tuple is supposed to be.

Reply:
Yeah, there's still a lot of messiness here that needs to be documented or implemented better for readability/maintainability purposes. I'm becoming noseblind to some of it because utils.py was even worse (https://github.com/pcdshub/iocmanager/blob/90c5bccac7e928a5052a8a594908ee36b82fadbd/utils.py#L685)
I'll take care of this one.